src/main/java/org/wildfly/security/http/HttpAuthenticator.java

```diff
@@ -0,0 +1,229 @@
+/*
+ * JBoss, Home of Professional Open Source.
+ * Copyright 2015 Red Hat, Inc., and individual contributors
+ * as indicated by the @author tags.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.wildfly.security.http;
+
+import static org.wildfly.security.http.HttpConstants.FORBIDDEN;
+import static org.wildfly.security.http.HttpConstants.OK;
+
+import java.util.List;
+import java.util.function.Supplier;
+
+import org.wildfly.security.auth.server.SecurityDomain;
+import org.wildfly.security.auth.server.SecurityIdentity;
+
+
+/**
+ * A HTTP based authenticator responsible for performing the authentication of the current request based on the policies of the
+ * associated {@link SecurityDomain}.
+ *
+ * @author <a href="mailto:darran.lofthouse@jboss.com">Darran Lofthouse</a>
+ */
+public class HttpAuthenticator {
+
+    private final Supplier<List<HttpServerAuthenticationMechanism>> mechanismSupplier;
+    private final HttpExchangeSpi httpExchangeSpi;
+    private final boolean required;
+    private final boolean ignoreOptionalFailures;
+    private volatile boolean authenticated = false;
+
+    private HttpAuthenticator(final Supplier<List<HttpServerAuthenticationMechanism>> mechanismSupplier, final HttpExchangeSpi httpExchangeSpi,
+            final boolean required, final boolean ignoreOptionalFailures) {
+        this.mechanismSupplier = mechanismSupplier;
+        this.httpExchangeSpi = httpExchangeSpi;
+        this.required = required;
+        this.ignoreOptionalFailures = ignoreOptionalFailures;
+    }
+
+    /**
+     * Perform authentication for the request.
+     *
+     * @return {@code true} if the call should be allowed to continue within the web server, {@code false} if the call should be
+     *         returning to the client.
+     * @throws HttpAuthenticationException
+     */
+    public boolean authenticate() throws HttpAuthenticationException {
+        return new AuthenticationExchange().authenticate();
+    }
+
+    private boolean isAuthenticated() {
+        return authenticated;
+    }
+
+    public static Builder builder() {
+        return new Builder();
+    }
+
+    private class AuthenticationExchange extends HttpServerExchange {
+
+        private volatile HttpServerAuthenticationMechanism currentMechanism;
+
+        private volatile int responseCode = -1;
+        private volatile boolean responseCodeAllowed = false;
```

Comment:
We probably will need to look at this one again but at this point I think the simple answer is we don't know the answer ;-) Depending on the architecture of the server we integrate with it possibly makes sense for us to allow sharing across threads at least from the perspective of ensuring visibility of updates to variables.